### PR TITLE
add expansion for promoted controbutors & count

### PR DESCRIPTION
**Pull request by sondrev**
Adds new field during expansion

### CODE REVIEW

File: expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java [hunk 1]
```diff
@@ -180,12 +187,35 @@ public String toString() {
...
+                                             .sorted(ExpandedResource::sortContributorByVerifiedFirst)
+                                             .limit(MAX_PROMOTED_CONTRIBUTORS).toList();
+
+                ArrayNode arrayNode = new ArrayNode(JsonNodeFactory.instance);
+                arrayNode.addAll(sortedContributors);
+
+                entityDescription.put(CONTRIBUTORS_COUNT, contributors.size());
+                entityDescription.put(CONTRIBUTORS_PROMOTED, arrayNode);
+            }
+        }
```
Comment (truhacevkir) [inline expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java:218]: maybe split it to multiple methods

**Review by truhacevkir — APPROVED** [review]

File: expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java [hunk 2]
```diff
@@ -180,12 +187,45 @@ public String toString() {
...
+    }
+
+    private static void injectContributorsPreview(ObjectNode json) {
+        var contributors = json.at(CONTRIBUTORS_PTR);
+        if (!contributors.isMissingNode() && contributors.isArray()) {
+            var entityDescription = (ObjectNode) json.at(ENTITY_DESCRIPTION_PTR);
+            if (!entityDescription.isMissingNode() && entityDescription.isObject()) {
+                List<JsonNode> contributorsList = new ArrayList<>();
```
Comment (monaullah) [inline expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java:216]: var

File: expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java [hunk 3]
```diff
@@ -180,12 +187,45 @@ public String toString() {
...
+                contributors.forEach(contributorsList::add);
+                var sortedContributors = contributorsList.stream()
+                                             .sorted(Comparator.comparingInt(c -> c.get(CONTRIBUTOR_SEQUENCE).asInt()))
+                                             .sorted(ExpandedResource::sortContributorByVerifiedFirst)
+                                             .limit(MAX_CONTRIBUTORS_PREVIEW).toList();
+
+                ArrayNode arrayNode = new ArrayNode(JsonNodeFactory.instance);
```
Comment (monaullah) [inline expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java:223]: var

File: expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java [hunk 4]
```diff
@@ -180,12 +187,45 @@ public String toString() {
...
+                arrayNode.addAll(sortedContributors);
+
+                entityDescription.put(CONTRIBUTORS_PREVIEW, arrayNode);
```
Comment (monaullah) [inline expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java:226]: `put` is deprecated, use `entityDescription.set(CONTRIBUTORS_PREVIEW, arrayNode);`

File: expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java [hunk 5]
```diff
@@ -180,12 +187,45 @@ public String toString() {
...
+
+    private static void injectContributorsPreview(ObjectNode json) {
+        var contributors = json.at(CONTRIBUTORS_PTR);
+        if (!contributors.isMissingNode() && contributors.isArray()) {
+            var entityDescription = (ObjectNode) json.at(ENTITY_DESCRIPTION_PTR);
+            if (!entityDescription.isMissingNode() && entityDescription.isObject()) {
+                List<JsonNode> contributorsList = new ArrayList<>();
+                contributors.forEach(contributorsList::add);
+                var sortedContributors = contributorsList.stream()
+                                             .sorted(Comparator.comparingInt(c -> c.get(CONTRIBUTOR_SEQUENCE).asInt()))
```
Comment (monaullah) [inline expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java:219]: rename `c` to `contributor`?

File: expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java [hunk 6]
```diff
@@ -180,12 +187,45 @@ public String toString() {
...
+        var contributors = json.at(CONTRIBUTORS_PTR);
+        if (!contributors.isMissingNode() && contributors.isArray()) {
+            var entityDescription = (ObjectNode) json.at(ENTITY_DESCRIPTION_PTR);
+            if (!entityDescription.isMissingNode() && entityDescription.isObject()) {
+                entityDescription.put(CONTRIBUTORS_COUNT, contributors.size());
+            }
+        }
+    }
+
+    private static void injectContributorsPreview(ObjectNode json) {
```
Comment (monaullah) [inline expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java:211]: This is quite a big method, refactor and extract to method to make more readable? Suggestion:

```
private static void injectContributorsPreview(ObjectNode json) {
        var contributors = json.at(CONTRIBUTORS_PTR);
        if (!contributors.isMissingNode() && contributors.isArray()) {
            var entityDescription = (ObjectNode) json.at(ENTITY_DESCRIPTION_PTR);
            if (!entityDescription.isMissingNode() && entityDescription.isObject()) {
                var sortedContributors = sortBySequenceAndVerificationLimitTen(contributors);
                var sortedContributorsArrayNode = new ArrayNode(JsonNodeFactory.instance).addAll(sortedContributors);

                entityDescription.set(CONTRIBUTORS_PREVIEW, sortedContributorsArrayNode);
            }
        }
    }

    private static List<JsonNode> sortBySequenceAndVerificationLimitTen(JsonNode contributors) {
        var contributorsList = new ArrayList<JsonNode>();
        contributors.forEach(contributorsList::add);
        return contributorsList.stream()
                   .sorted(Comparator.comparingInt(contributor -> contributor.get(CONTRIBUTOR_SEQUENCE).asInt()))
                   .sorted(ExpandedResource::sortContributorByVerifiedFirst)
                   .limit(MAX_CONTRIBUTORS_PREVIEW).toList();
    }
```

File: expansion/src/test/java/no/unit/nva/publication/indexing/ExpandedResourceTest.java [hunk 7]
```diff
@@ -182,6 +187,37 @@ void shouldReturnIndexDocumentWithValidContributorAffiliationCountryCode() throw
         assertThat(actualCountryCode, is(not(nullValue())));
     }
 
+    @Test
+    void shouldReturnIndexDocumentWithContributorsPreviewAndCount() throws Exception {
```
Comment (monaullah) [inline expansion/src/test/java/no/unit/nva/publication/indexing/ExpandedResourceTest.java:191]: Test limit? Test data setup with more than 10 contributors?

**Review by monaullah — CHANGES_REQUESTED** [review]

**Review by brinxmat — APPROVED** [review]
Approving to move on, but see other reviewer's comments

**Pull request merged**